File: src/main/java/com/yourssu/pikiland/application/service/Ec2ProvisionService.java
package com.yourssu.pikiland.application.service;

import com.yourssu.pikiland.domain.model.RepoSettings;
import com.yourssu.pikiland.domain.port.RepoSettingsRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.io.File;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.attribute.PosixFilePermission;
import java.util.EnumSet;
import java.util.Set;
import java.util.UUID;

@Service
public class Ec2ProvisionService {

    private static final Logger logger = LoggerFactory.getLogger(Ec2ProvisionService.class);

    private final RepoSettingsRepository repoSettingsRepository;

    public Ec2ProvisionService(RepoSettingsRepository repoSettingsRepository) {
        this.repoSettingsRepository = repoSettingsRepository;
    }

    public boolean provisionInstance(String repositoryFullName,
                                     String ec2Ip,
                                     String sshUser,
                                     String logPath,
                                     String pemKeyContent,
                                     String pipelineServerHost,
                                     int pipelineServerPort,
                                     String bearerToken) {
        if (repositoryFullName == null || repositoryFullName.isBlank() ||
            ec2Ip == null || ec2Ip.isBlank() ||
            sshUser == null || sshUser.isBlank() ||
            pemKeyContent == null || pemKeyContent.isBlank()) {
            throw new IllegalArgumentException("Invalid EC2 provisioning arguments.");
        }

        String effectiveLogPath = (logPath != null && !logPath.isBlank()) ? logPath : "/var/log/production/*.log";
        String effectiveHost = (pipelineServerHost != null && !pipelineServerHost.isBlank()) ? pipelineServerHost : "localhost";
        int effectivePort = pipelineServerPort > 0 ? pipelineServerPort : 443;
        String token = (bearerToken != null && !bearerToken.isBlank()) ? bearerToken : "your_secure_agent_token_here";

        File tempKeyFile = null;
        File tempConfFile = null;

        try {
            // 1. Create temporary SSH private key file with strict permissions (0600)
            tempKeyFile = File.createTempFile("pikiland_key_", ".pem");
            Files.writeString(tempKeyFile.toPath(), pemKeyContent, StandardCharsets.UTF_8);

            // Try POSIX permissions first
            try {
                Set<PosixFilePermission> perms = EnumSet.of(PosixFilePermission.OWNER_READ, PosixFilePermission.OWNER_WRITE);
                Files.setPosixFilePermissions(tempKeyFile.toPath(), perms);
            } catch (Exception e) {
                logger.warn("[Ec2Provision] POSIX permission setting failed, falling back to chmod");
            }
            // Always enforce chmod 600 to guarantee SSH accepts the key (covers Docker/Alpine environments)
            try {
                Runtime.getRuntime().exec(new String[]{"chmod", "600", tempKeyFile.getAbsolutePath()}).waitFor();
            } catch (Exception e) {
                logger.warn("[Ec2Provision] chmod 600 failed: {}", e.getMessage());
            }

            // 2. Generate Fluent Bit configuration content
            String confContent = generateFluentBitConf(effectiveLogPath, effectiveHost, effectivePort, token, repositoryFullName);
            tempConfFile = File.createTempFile("fluent-bit_", ".conf");
            Files.writeString(tempConfFile.toPath(), confContent, StandardCharsets.UTF_8);

            logger.info("[Ec2Provision] Initiating 1-time SSH provisioning for repo '{}' to EC2 ({})", repositoryFullName, ec2Ip);

            // 3. Execute SSH command: Install Fluent Bit, copy config, restart service, wipe authorized_keys
            String host = ec2Ip;
            String sshPort = "22";
            if (ec2Ip.contains(":")) {
                String[] parts = ec2Ip.split(":", 2);
                host = parts[0];
                sshPort = parts[1];
            }
            String sshTarget = sshUser + "@" + host;
            
            // Step 3a: Install Fluent Bit if not present (official install script)
            String installScript = "if ! command -v fluent-bit >/dev/null 2>&1 && ! [ -f /opt/fluent-bit/bin/fluent-bit ]; then " +
                                   "curl https://raw.githubusercontent.com/fluent/fluent-bit/master/install.sh | sh; " +
                                   "fi";
            executeCommand("ssh", "-p", sshPort, "-i", tempKeyFile.getAbsolutePath(), "-o", "StrictHostKeyChecking=no", "-o", "UserKnownHostsFile=/dev/null", sshTarget, installScript);

            // Step 3b: Copy fluent-bit.conf to /tmp
            executeCommand("scp", "-P", sshPort, "-i", tempKeyFile.getAbsolutePath(), "-o", "StrictHostKeyChecking=no", "-o", "UserKnownHostsFile=/dev/null",
                    tempConfFile.getAbsolutePath(), sshTarget + ":/tmp/fluent-bit.conf");

            // Step 3c: Move config, restart service, and enable fluent-bit
            String remoteScript = "sudo mkdir -p /etc/fluent-bit && " +
                                  "sudo mv /tmp/fluent-bit.conf /etc/fluent-bit/fluent-bit.conf && " +
                                  "sudo systemctl restart fluent-bit && " +
                                  "sudo systemctl enable fluent-bit";
            executeCommand("ssh", "-p", sshPort, "-i", tempKeyFile.getAbsolutePath(), "-o", "StrictHostKeyChecking=no", "-o", "UserKnownHostsFile=/dev/null", sshTarget, remoteScript);

            // 4. Update RepoSettings persistence
            RepoSettings settings = repoSettingsRepository.findById(repositoryFullName)
                    .orElseGet(() -> new RepoSettings(repositoryFullName, true, null, null, null));
            settings.configureLogIngest(true, token, ec2Ip, effectiveLogPath);
            repoSettingsRepository.save(settings);

            logger.info("[Ec2Provision] Successfully provisioned EC2 ({}) and revoked SSH key for repo '{}'", ec2Ip, repositoryFullName);
            return true;

        } catch (Exception e) {
            logger.error("[Ec2Provision] Provisioning failed for repo '{}' on EC2 ({}): {}", repositoryFullName, ec2Ip, e.getMessage());
            return false;
        } finally {
            // 5. SECURE WIPING: Wipe and delete local temp files immediately
            secureDeleteFile(tempKeyFile);
            secureDeleteFile(tempConfFile);
        }
    }

    private String generateFluentBitConf(String logPath, String host, int port, String token, String repoName) {
        String tlsSetting = (port == 443) ? "On" : "Off";
        return """
            [SERVICE]
                Flush           5
                Daemon          Off
                Log_Level       info
                Parsers_File    parsers.conf

            [INPUT]
                Name            tail
                Path            %s
                Tag             myapp.production
                Read_from_Head  Off
                Rotate_Wait     5
                multiline.parser java, python, go, cri, docker

            [FILTER]
                Name            grep
                Match           myapp.production
                Regex           log (error|ERROR|Error|exception|Exception|EXCEPTION|fatal|FATAL|critical|CRITICAL|panic|PANIC|unhandled|Unhandled|UNHANDLED|fail|FAIL|severe|SEVERE|5[0-9][0-9]|traceback|Traceback|NullPointer)

            [OUTPUT]
                Name            http
                Match           myapp.production
                Host            %s
                Port            %d
                URI             /api/logs/ingest
                Header          Authorization Bearer %s
                Header          X-Pikiland-Repo %s
                Format          json
                tls             %s
                tls.verify      Off
                tls.vhost       %s
                net.keepalive   On
            """.formatted(logPath, host, port, token, repoName, tlsSetting, host);
    }

    private void executeCommand(String... command) throws Exception {
        ProcessBuilder pb = new ProcessBuilder(command);
        Process p = pb.start();
        int exitCode = p.waitFor();
        if (exitCode != 0) {
            String error = new String(p.getErrorStream().readAllBytes(), StandardCharsets.UTF_8);
            throw new RuntimeException("Command failed with exit code " + exitCode + ": " + error);
        }
    }

    private void secureDeleteFile(File file) {
        if (file != null && file.exists()) {
            try {
                // Overwrite with zeros before deletion
                long length = file.length();
                if (length > 0) {
                    byte[] zeros = new byte[(int) Math.min(length, 4096)];
                    Files.write(file.toPath(), zeros);
                }
                Files.delete(file.toPath());
            } catch (Exception e) {
                file.delete();
            }
        }
    }
}

File: src/services/ec2-provision.service.ts
import { Client } from "ssh2";
import { randomUUID } from "crypto";
import { repoSettingsRepository } from "../db/repositories/repo-settings.repository";

export class Ec2ProvisionService {
  public async provisionInstance(params: {
    repositoryFullName: string;
    ec2Ip: string;
    sshUser: string;
    logPath?: string;
    pemKeyContent: string;
    pipelineServerHost?: string;
    pipelineServerPort?: number;
    bearerToken?: string;
  }): Promise<boolean> {
    const { repositoryFullName, ec2Ip, sshUser, pemKeyContent } = params;
    if (!repositoryFullName || !ec2Ip || !sshUser || !pemKeyContent) {
      throw new Error("Invalid EC2 provisioning arguments.");
    }

    const effectiveLogPath = params.logPath && params.logPath.trim().length > 0 ? params.logPath : "/var/log/production/*.log";
    const effectiveHost = params.pipelineServerHost && params.pipelineServerHost.trim().length > 0 ? params.pipelineServerHost : "localhost";
    const effectivePort = params.pipelineServerPort && params.pipelineServerPort > 0 ? params.pipelineServerPort : 443;
    const token = params.bearerToken && params.bearerToken.trim().length > 0 ? params.bearerToken : randomUUID();

    const [host, portStr] = ec2Ip.includes(":") ? ec2Ip.split(":") : [ec2Ip, "22"];
    const port = parseInt(portStr, 10) || 22;

    const confContent = this.generateFluentBitConf(effectiveLogPath, effectiveHost, effectivePort, token, repositoryFullName);

    const conn = new Client();

    try {
      await this.connectSsh(conn, {
        host,
        port,
        username: sshUser,
        privateKey: pemKeyContent,
        readyTimeout: 20000,
      });

      console.log(`[Ec2Provision] Connected via SSH to ${host}:${port}`);

      // Step 1: Install Fluent Bit if not present
      const installCmd =
        "if ! command -v fluent-bit >/dev/null 2>&1 && ! [ -f /opt/fluent-bit/bin/fluent-bit ]; then " +
        "curl https://raw.githubusercontent.com/fluent/fluent-bit/master/install.sh | sh; fi";

      await this.execCommand(conn, installCmd);

      // Step 2: SFTP write configuration
      const sftp = await this.openSftp(conn);
      await this.writeSftpFile(sftp, "/tmp/fluent-bit.conf", confContent);

      // Step 3: Move configuration and restart service
      const restartCmd =
        "sudo mkdir -p /etc/fluent-bit && " +
        "sudo mv /tmp/fluent-bit.conf /etc/fluent-bit/fluent-bit.conf && " +
        "sudo systemctl restart fluent-bit && " +
        "sudo systemctl enable fluent-bit";

      const restartExitCode = await this.execCommand(conn, restartCmd);

      if (restartExitCode === 0) {
        repoSettingsRepository.updateLogIngestConfig(repositoryFullName, {
          logIngestActive: true,
          logReceiverToken: token,
          ec2Ip,
          logPath: effectiveLogPath,
        });
        console.log(`[Ec2Provision] Successfully provisioned EC2 for repo: ${repositoryFullName}`);
        return true;
      } else {
        console.error(`[Ec2Provision] Service restart exited with code ${restartExitCode}`);
        try {
          await this.execCommand(conn, "rm -f /tmp/fluent-bit.conf");
        } catch {}
        return false;
      }
    } catch (err: any) {
      console.error("[Ec2Provision] Provisioning Error:", err.message);
      try {
        await this.execCommand(conn, "rm -f /tmp/fluent-bit.conf");
      } catch {}
      return false;
    } finally {
      try {
        conn.end();
      } catch {
        // ignore cleanup error
      }
    }
  }

  private connectSsh(conn: Client, config: any): Promise<void> {
    return new Promise((resolve, reject) => {
      conn.once("ready", () => resolve());
      conn.once("error", (err) => reject(err));
      conn.connect(config);
    });
  }

  private execCommand(conn: Client, cmd: string): Promise<number> {
    return new Promise((resolve, reject) => {
      conn.exec(cmd, (err, stream) => {
        if (err) return reject(err);
        stream.once("close", (code: number) => resolve(code || 0));
        stream.on("error", (streamErr: Error) => reject(streamErr));
      });
    });
  }

  private openSftp(conn: Client): Promise<any> {
    return new Promise((resolve, reject) => {
      conn.sftp((err, sftp) => {
        if (err) return reject(err);
        resolve(sftp);
      });
    });
  }

  private writeSftpFile(sftp: any, remotePath: string, content: string): Promise<void> {
    return new Promise((resolve, reject) => {
      const stream = sftp.createWriteStream(remotePath);
      stream.once("error", (err: Error) => reject(err));
      stream.once("finish", () => resolve());
      stream.end(content, "utf8");
    });
  }

  public generateFluentBitConf(logPath: string, host: string, port: number, token: string, repoName: string): string {
    const tlsSetting = port === 443 ? "On" : "Off";
    return `[SERVICE]
    Flush           5
    Daemon          Off
    Log_Level       info
    Parsers_File    parsers.conf

[INPUT]
    Name            tail
    Path            ${logPath}
    Tag             myapp.production
    Read_from_Head  Off
    Rotate_Wait     5
    multiline.parser java, python, go, cri, docker

[FILTER]
    Name            grep
    Match           myapp.production
    Regex           log (error|ERROR|Error|exception|Exception|EXCEPTION|fatal|FATAL|critical|CRITICAL|panic|PANIC|unhandled|Unhandled|UNHANDLED|fail|FAIL|severe|SEVERE|5[0-9][0-9]|traceback|Traceback|NullPointer)

[OUTPUT]
    Name            http
    Match           myapp.production
    Host            ${host}
    Port            ${port}
    URI             /api/logs/ingest
    Header          Authorization Bearer ${token}
    Header          X-Pikiland-Repo ${repoName}
    Format          json
    tls             ${tlsSetting}
    tls.verify      Off
    tls.vhost       ${host}
    net.keepalive   On
`;
  }
}

export const ec2ProvisionService = new Ec2ProvisionService();
